File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/domain/policies/TopUpPaymentRequestPolicyTest.java
package it.giannibombelli.wsc2026.giftcard.domain.policies;

import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardTopUpRequested;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.createGiftCard;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * La policy è deliberatamente uno stub; il test verifica solo che venga invocata.
 */
class TopUpPaymentRequestPolicyTest {

    @Test
    void evaluate_withNullEvent_throwsIllegalArgumentException() {
        TopUpPaymentRequestPolicy policy = new TopUpPaymentRequestPolicy();

        assertThatThrownBy(() -> policy.evaluate(null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void evaluate_returnsRequestPayment() {
        GiftCard card = createGiftCard();
        GiftCardTopUpRequested event = card.requestTopUp(new Money(new BigDecimal("25.50")));

        TopUpPaymentRequestPolicy policy = new TopUpPaymentRequestPolicy();

        var cmd = policy.evaluate(event);

        org.assertj.core.api.Assertions.assertThat(cmd.clientReference()).isEqualTo(new ClientReference(event.aggregateId().value()));
        org.assertj.core.api.Assertions.assertThat(cmd.amount()).isEqualTo(event.requestedAmount());
    }
}

File: typescript/tests/giftcard/application/policies/topUpPaymentRequestPolicy.test.ts
import { describe, expect, it } from 'vitest';
import { Money } from '@/common/domain/primitive/money.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { TopUpPaymentRequestPolicy } from '@/giftcard/application/policies/topUpPaymentRequestPolicy.js';
import { GiftCardAggregateFactory } from '../../../testsupport/giftcard/aggregateFactory.js';

describe('TopUpPaymentRequestPolicy', () => {
  it('evaluate with null event throws', () => {
    const policy = new TopUpPaymentRequestPolicy();

    expect(() => policy.evaluate(null as unknown as Parameters<typeof policy.evaluate>[0])).toThrow();
  });

  it('evaluate returns RequestPayment', () => {
    const card = GiftCardAggregateFactory.createGiftCard();
    const event = card.requestTopUp(new Money(25.5));
    const policy = new TopUpPaymentRequestPolicy();

    const cmd = policy.evaluate(event);

    expect(cmd.clientReference.toString()).toEqual(event.aggregateId.value.value);
    expect(cmd.amount).toEqual(event.requestedAmount);
    expect(cmd.aggregateId).toBeInstanceOf(PaymentId);
  });
});
